### DISCUSSION TITLE
Why is $args handled differently in these two aliases?

### DISCUSSION THREAD

**Discussion by CmdQ** [xonsh/xonsh#5027] [Q&A] [answered]
So I'm converting my Git alias collection to Xonsh. Take a look at these two:

1. `aliases["gf"] = "git fetch"`
2. `aliases["grbm"] = "git rebase @(git_default_main_branch())`

I extend the first one successfully with more arguments, like `-h` for instance and it shows me the help of fetch. Why doesn't it work with the second one?

Doing `grbm -i` does _not_ bring up the editor but immediately rebases (also `-h` has no effect). Manually changing the second alias to `aliases["grbm"] = "git rebase @(git_default_main_branch()) @($args)` makes it behave like I expect.  
But why do I need to handle that one differently?

And how do I know that I _need_ to?

**Comment by anki-code** [comment]
It's this question/proposal should be an issue. On my opinion the root of difference is in that the first example is "simple" subprocess alias but the second one is ExecAlias that require interpretation as xonsh code. I suppose when ExecAlias was invented the inventor just didn't think about consistency with the first example and hadn't add adding the args at the end by default. 
But maybe it was done for more careful control on arguments without hiding behavior.

**Reply by CmdQ** [reply]
You mean I should write an issue?

**Reply by anki-code** [reply]
I think the core team should discuss this difference before releasing 1.0. But I think there will be no changes because as I wrote above these two cases are different by design and if we add args to ExecAlias by default we lose the power of args management in ExecAlias. When you choose ExecAlias you may want to do `aliases['example'] = "command @($arg1) -opt @($arg2) -opt @($args[2:])"` and it's more powerful than just adding args at the end of the command.

**Reply by CmdQ** [reply]
OK, thanks for the explanation.

I get both points. Losing the power in `@($args[2:])` would be bad. But it's also confusing to newcomers that expect things to work the same if it looks the same.  
That one is simple and the other is a `ExecAlias` is in no way apparent—and you also shouldn't need to know about `ExecAlias`.

Maybe a heuristic along the lines of “`$args` not used in the alias ⇒ slap it at the end”. And if you really wanted an alias that ignores everything in there you might do something like `aliases["ignore"] = "echo @('ignoring') # $args"`.

At the very least, it should say in the documentation.

**Comment by anki-code** [comment, marked as answer]
Xonsh tutorial has the description in [Alias section](https://xon.sh/tutorial.html#aliases) ([code](https://github.com/xonsh/xonsh/blob/main/docs/tutorial.rst)):

> **If the string is representing a block of xonsh code, the alias will be registered as an ExecAlias, which is a callable alias**. This block of code will then be executed whenever the alias is run. **The arguments are available in the list $args** or by the index in $arg<n> environment variables.
> aliases['answer'] = 'echo @(21+21)'
> aliases['piu'] = 'pip install -U @($args)'
> aliases['cdls'] = 'cd $arg0 && ls'

Feel free to make a PR with update this section!

**Reply by CmdQ** [reply]
OK. One more question: _Why_ does it become `ExecAlias` mode? Because I used `@(…)`, right?

**Reply by anki-code** [reply]
yep
